### PR TITLE
feat(Table): Allow authors to specify frozen columns

### LINKED ISSUE #715: Table: Allow authors to freeze column(s)
## Description
Allow authors to specify column(s) to be frozen (always visible) in Table components.

### PR DESCRIPTION
**Pull request by breity**
## Changes
Adds a checkbox for each column in Table component authoring that specifies whether the column in frozen or not.

Tabulator allows any number of frozen columns on each side of the table. Frozen columns must be next to each other. Once a non-frozen column is reached, Tabulator stops building the frozen list for that side of the table. Total number of frozen columns must be less than total number of columns.

Closes #715.

## Test
- When authoring a Table component Select the "Freeze" checkbox for any columns in the table.
- Verify in VLE (and grading) that any columns you checked are frozen in Tabulator (if they're next to each other).
- Verify that you can't select all of the columns to be frozen. (If you select n-1 frozen columns, the last remaining column should have it's 'Frozen' checkbox disabled.)

### CODE REVIEW

File: src/assets/wise5/components/table/table-authoring/table-authoring.component.ts [hunk 1]
```diff
@@ -308,21 +313,62 @@ export class TableAuthoring extends ComponentAuthoring {
...
+      return false;
+    }
+    const maxFrozen = firstRow.length - 1;
+    for (const key in firstRow) {
+      if (firstRow[key].frozen) {
+        count++;
+      }
+    }
+    return count === maxFrozen;
+  }
```
Comment (hirokiterashima) [inline src/assets/wise5/components/table/table-authoring/table-authoring.component.ts]: Move variable declarations just before you need them
```suggestion
  private isfrozenColumnsLimitReached(): boolean {
    const firstRow = this.authoringComponentContent.tableData[0];
    if (firstRow == null) {
      return false;
    }
    let count = 0;
    for (const key in firstRow) {
      if (firstRow[key].frozen) {
        count++;
      }
    }
    const maxFrozen = firstRow.length - 1;
    return count === maxFrozen;
  }
```

File: src/assets/wise5/components/table/table-authoring/table-authoring.component.html [hunk 2]
```diff
@@ -153,6 +153,17 @@
                 (ngModelChange)="columnSizeChanged(cellIndex)"/>
           </mat-form-field>
         </div>
+        <div>
+          <mat-checkbox color="primary"
+              [(ngModel)]="frozenColumns[cellIndex]"
+              (ngModelChange)="frozenColumnsChanged(cellIndex)"
+              [disabled]="frozenColumnsLimitReached && !frozenColumns[cellIndex]"
+              matTooltip="You can freeze any number of columns to the beginning or end of the table. Frozen columns must be next to each other."
```
Comment (hirokiterashima) [inline src/assets/wise5/components/table/table-authoring/table-authoring.component.html]: Update tooltip text. Frozen columns do not need to be next to each other.

**Review by hirokiterashima — CHANGES_REQUESTED** [review]
Functionality works well. Please see comments inline.

**Comment by breity** [comment]
@hirokiterashima Thanks, I updated with the requested changes.

**Review by hirokiterashima — APPROVED** [review]
Looks good. Can you make this function private before you merge?
```
  setColumnCellSize(column: number, size: number): void {
```

**Pull request merged**